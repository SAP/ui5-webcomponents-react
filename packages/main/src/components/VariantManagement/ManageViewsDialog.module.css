.manageViewsDialog {
  width: 100%;

  &::part(content),
  &::part(header) {
    padding: 0;
  }

  &::part(footer) {
    padding: 0;
    border-block-start: none;
  }
}

@media (min-width: 1024px) {
  .manageViewsDialog {
    width: 70vw;
  }
}

.headerText {
  margin: 0;
  text-align: center;
  align-self: start;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
  max-width: 100%;
  display: inline-block;
  padding-inline-start: 1rem;
  font-family: var(--_ui5wcr_popup_header_font_family);
  font-size: 1rem;
  min-height: var(--_ui5wcr_popup_default_header_height);
  max-height: var(--_ui5wcr_popup_default_header_height);
  line-height: var(--_ui5wcr_popup_default_header_height);
}

.search {
  width: calc(100% - 2rem);
  margin-block-end: 0.5rem;
}

.inputIcon {
  cursor: pointer;
  color: var(--sapContent_IconColor);
}
