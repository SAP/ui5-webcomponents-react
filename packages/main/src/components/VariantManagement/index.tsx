import '@ui5/webcomponents-fiori/dist/illustrations/UnableToLoad.js';
import '@ui5/webcomponents-icons/dist/decline.js';
import '@ui5/webcomponents-icons/dist/navigation-down-arrow.js';
import '@ui5/webcomponents-icons/dist/search.js';
import { useI18nBundle } from '@ui5/webcomponents-react-base/dist/hooks';
import { ThemingParameters } from '@ui5/webcomponents-react-base/dist/ThemingParameters';
import { enrichEventWithDetails } from '@ui5/webcomponents-react-base/dist/Utils';
import {
  CANCEL,
  MANAGE,
  MY_VIEWS,
  RESET,
  SAVE,
  SAVE_AS,
  SEARCH,
  SEARCH_VARIANT,
  SELECT_VIEW
} from '@ui5/webcomponents-react/dist/assets/i18n/i18n-defaults';
import { Bar } from '@ui5/webcomponents-react/dist/Bar';
import { Button } from '@ui5/webcomponents-react/dist/Button';
import { ButtonDesign } from '@ui5/webcomponents-react/dist/ButtonDesign';
import { FlexBox } from '@ui5/webcomponents-react/dist/FlexBox';
import { Icon } from '@ui5/webcomponents-react/dist/Icon';
import { IllustratedMessage } from '@ui5/webcomponents-react/dist/IllustratedMessage';
import { IllustrationMessageType } from '@ui5/webcomponents-react/dist/IllustrationMessageType';
import { Input } from '@ui5/webcomponents-react/dist/Input';
import { List } from '@ui5/webcomponents-react/dist/List';
import { ListMode } from '@ui5/webcomponents-react/dist/ListMode';
import { PopoverPlacementType } from '@ui5/webcomponents-react/dist/PopoverPlacementType';
import { ResponsivePopover, ResponsivePopoverDomRef } from '@ui5/webcomponents-react/dist/ResponsivePopover';
import { Title } from '@ui5/webcomponents-react/dist/Title';
import { TitleLevel } from '@ui5/webcomponents-react/dist/TitleLevel';
import { SelectedVariant, VariantManagementContext } from '@ui5/webcomponents-react/dist/VariantManagementContext';
import { CommonProps } from '@ui5/webcomponents-react/interfaces/CommonProps';
import React, {
  Children,
  cloneElement,
  ComponentElement,
  forwardRef,
  isValidElement,
  ReactNode,
  Ref,
  useCallback,
  useEffect,
  useRef,
  useState
} from 'react';
import { createPortal } from 'react-dom';
import { createUseStyles } from 'react-jss';
import { Ui5CustomEvent } from '../../interfaces/Ui5CustomEvent';
import { stopPropagation } from '../../internal/stopPropagation';
import { ManageViewsDialog } from './ManageViewsDialog';
import { SaveViewDialog } from './SaveViewDialog';
import { VariantItemPropTypes } from './VariantItem';
import clsx from 'clsx';

interface UpdatedVariant extends SelectedVariant {
  prevVariant?: VariantItemPropTypes;
}

export interface VariantManagementPropTypes extends Omit<CommonProps, 'onSelect'> {
  /**
   * Variant items displayed by the VariantManagement component.
   *
   * __Note:__ Although this prop accepts all HTML Elements, it is strongly recommended that you only use `VariantItem` in order to preserve the intended design.
   */
  children?: ReactNode | ReactNode[];
  /**
   * Determines on which side the VariantManagement popover is placed at.
   */
  placement?: PopoverPlacementType | keyof typeof PopoverPlacementType;
  /**
   * Describes the title of the VariantManagement popover.
   *
   * __Note:__ If not set, the default title is used.
   */
  titleText?: string;
  /**
   * Defines whether the VariantManagement should be closed if an item was selected.
   */
  closeOnItemSelect?: boolean;
  /**
   * Describes the `HTML Title` level of the variants.
   */
  level?: TitleLevel | keyof typeof TitleLevel;
  /**
   * Defines whether the VariantManagement is disabled.
   */
  disabled?: boolean;
  /**
   * Fired after a variant has been selected.
   */
  onSelect?: (
    event: Ui5CustomEvent<
      HTMLElement,
      {
        selectedVariant: SelectedVariant;
        selectedItems: unknown[];
        previouslySelectedItems: unknown[];
      }
    >
  ) => void;
  /**
   * Indicator for modified but not saved variants.
   *
   * __Note:__ You can change the indicator by setting `dirtyStateText`.
   */
  dirtyState?: boolean;
  /**
   * Text for the dirty state indicator.
   */
  dirtyStateText?: string;
  /**
   * Indicates that the 'Favorites' feature is used. Only variants marked as favorites will be displayed in the variant list.
   */
  showOnlyFavorites?: boolean;
  /**
   * Indicates that set as default is visible in the Save View and the Manage Views dialogs.
   */
  hideSetAsDefault?: boolean;
  /**
   * Indicates that the Public indicator is visible in the Save View and the Manage Views dialogs.
   */
  hideShare?: boolean;
  /**
   * Indicates that Apply Automatically is visible in the Save View and the Manage Views dialogs.
   */
  hideApplyAutomatically?: boolean;
  /**
   * Indicates that the Save View dialog button is visible.
   */
  hideSaveAs?: boolean;
  /**
   * Indicates that the Manage Views dialog button is visible.
   */
  hideManageVariants?: boolean;
  /**
   * Displays the cancel button in the popover.
   */
  showCancelButton?: boolean;
  /**
   * Indicates that the control is in error state. If set to true error message will be displayed whenever the variant is opened.
   */
  inErrorState?: boolean;
  /**
   * Defines where modals are rendered into via `React.createPortal`.
   *
   * Defaults to: `document.body`
   */
  portalContainer?: Element;
  /**
   * The event is fired when the "Save" button is clicked inside the Save View dialog.
   */
  onSaveAs?: (e: CustomEvent<SelectedVariant>) => void;
  /**
   * The event is fired when the "Save" button is clicked inside the Manage Views dialog.
   */
  onSaveManageViews?: (
    e: CustomEvent<{
      deletedVariants: VariantItemPropTypes[];
      prevVariants: VariantItemPropTypes[];
      updatedVariants: UpdatedVariant[];
      variants: SelectedVariant[];
    }>
  ) => void;
  /**
   * The event is fired when the "Save" button is clicked in the `VariantManagement` popover.
   *
   * __Note:__ The save button is only displayed if the `VariantManagement` is in `dirtyState` and the selected variant is not in `readOnly` mode.
   */
  onSave?: (e: CustomEvent<SelectedVariant>) => void;
}

const styles = {
  container: {
    display: 'flex',
    alignItems: 'center',
    textAlign: 'center'
  },
  title: {
    cursor: 'pointer',
    color: ThemingParameters.sapButton_TextColor,
    '&:hover': {
      color: ThemingParameters.sapButton_Hover_TextColor
    }
  },
  disabled: {
    '& $title': {
      color: ThemingParameters.sapGroup_TitleTextColor,
      cursor: 'default',
      '&:hover': {
        color: 'ThemingParameters.sapGroup_TitleTextColor'
      }
    }
  },
  dirtyState: {
    color: ThemingParameters.sapGroup_TitleTextColor,
    padding: '0 0.125rem',
    fontWeight: 'bold',
    font: ThemingParameters.sapFontFamily,
    fontSize: ThemingParameters.sapFontSize,
    flexGrow: 1
  },
  dirtyStateText: {
    fontSize: ThemingParameters.sapFontSmallSize,
    fontWeight: 'normal'
  },
  footer: {
    margin: '0.4375rem 1rem 0.4325rem auto'
  },
  inputIcon: { cursor: 'pointer', color: ThemingParameters.sapContent_IconColor },
  searchInput: { padding: '0.25rem 0.5rem 0.25rem 0.25rem' },
  popover: { minWidth: '25rem' }
};

const useStyles = createUseStyles(styles, { name: 'VariantManagement' });
/**
 * The `VariantManagement` component can be used to manage variants, such as FilterBar variants or AnalyticalTable variants.
 */
const VariantManagement = forwardRef((props: VariantManagementPropTypes, ref: Ref<HTMLDivElement>) => {
  const i18nBundle = useI18nBundle('@ui5/webcomponents-react');
  const {
    titleText = i18nBundle.getText(MY_VIEWS),
    className,
    style,
    tooltip,
    placement,
    level,
    onSelect,
    closeOnItemSelect,
    disabled,
    onSaveAs,
    onSaveManageViews,
    showOnlyFavorites,
    inErrorState,
    hideShare,
    children,
    hideManageVariants,
    hideApplyAutomatically,
    hideSetAsDefault,
    hideSaveAs,
    dirtyStateText,
    dirtyState,
    showCancelButton,
    onSave,
    portalContainer,
    ...rest
  } = props;

  const classes = useStyles();
  const popoverRef = useRef<ResponsivePopoverDomRef>(null);

  const [safeChildren, setSafeChildren] = useState(Children.toArray(children));
  const [showInput, setShowInput] = useState(safeChildren.length > 9);

  useEffect(() => {
    setSafeChildren(Children.toArray(children));
  }, [children]);

  useEffect(() => {
    if (safeChildren.length > 9) {
      setShowInput(true);
    } else {
      setShowInput(false);
    }
  }, [safeChildren.length]);

  const [manageViewsDialogOpen, setManageViewsDialogOpen] = useState(false);
  const [saveAsDialogOpen, setSaveAsDialogOpen] = useState(false);
  const [selectedVariant, setSelectedVariant] = useState<SelectedVariant | undefined>(() => {
    const currentSelectedVariant = safeChildren.find(
      (item) => isValidElement(item) && item.props.selected
    ) as ComponentElement<any, any>;
    if (currentSelectedVariant) {
      return { ...currentSelectedVariant.props, variantItem: currentSelectedVariant.ref };
    }
  });

  const handleClose = () => {
    popoverRef.current.close();
  };

  const handleManageClick = () => {
    setManageViewsDialogOpen(true);
  };
  const handleManageClose = () => {
    setManageViewsDialogOpen(false);
  };
  const handleOpenSaveAsDialog = () => {
    setSaveAsDialogOpen(true);
  };
  const handleSaveAsClose = () => {
    setSaveAsDialogOpen(false);
  };

  const handleSave = (e) => {
    if (typeof onSave === 'function') {
      onSave(enrichEventWithDetails(e, selectedVariant as Record<string, any>));
    }
  };

  const handleSaveView = (e, selectedVariant) => {
    if (typeof onSaveAs === 'function') {
      onSaveAs(enrichEventWithDetails(e, selectedVariant));
    }
    handleSaveAsClose();
  };

  const handleSaveManageViews = (e, payload) => {
    const { defaultView, updatedRows, deletedRows } = payload;
    let callbackProperties = { deletedVariants: [], prevVariants: [], updatedVariants: [], variants: [] };
    if (typeof onSaveManageViews === 'function') {
      onSaveManageViews(enrichEventWithDetails(e, callbackProperties));
    }
    setSafeChildren((prev) =>
      prev
        .map((child: ComponentElement<any, any>) => {
          let updatedProps: Omit<SelectedVariant, 'children' | 'variantItem'> = {};
          const currentVariant = popoverRef.current.querySelector(`ui5-li[data-text="${child.props.children}"]`);
          callbackProperties.prevVariants.push(child.props);
          if (defaultView) {
            if (defaultView === child.props.children) {
              updatedProps.isDefault = true;
            } else if (child.props.isDefault) {
              updatedProps.isDefault = false;
            }
          }
          if (Object.keys(updatedRows).includes(child.props.children)) {
            const { currentVariant, ...rest } = updatedRows[child.props.children];
            updatedProps = { ...updatedProps, ...rest };
          }
          if (deletedRows.has(child.props.children)) {
            callbackProperties.deletedVariants.push(child.props);
            return false;
          }
          if (Object.keys(updatedProps).length > 0) {
            callbackProperties.updatedVariants.push({
              ...child.props,
              ...updatedProps,
              variantItem: currentVariant,
              prevVariant: { ...child.props }
            });
          }
          callbackProperties.variants.push({ ...child.props, ...updatedProps, variantItem: currentVariant });
          return cloneElement(child, updatedProps);
        })
        .filter(Boolean)
    );
    handleManageClose();
  };

  const handleOpenVariantManagement = useCallback(
    (e) => {
      popoverRef.current.showAt(e.target);
    },
    [popoverRef]
  );

  const cancelText = i18nBundle.getText(CANCEL);
  const searchText = i18nBundle.getText(SEARCH);
  const saveAsText = i18nBundle.getText(SAVE_AS);
  const manageText = i18nBundle.getText(MANAGE);
  const saveText = i18nBundle.getText(SAVE);
  const a11ySearchText = i18nBundle.getText(SEARCH_VARIANT);
  const selectViewText = i18nBundle.getText(SELECT_VIEW);
  const resetIconTitleText = i18nBundle.getText(RESET);

  const variantManagementClasses = clsx(classes.container, disabled && classes.disabled, className);

  const dirtyStateClasses = clsx(classes.dirtyState, dirtyStateText !== '*' && classes.dirtyStateText);

  const selectVariantEventRef = useRef();
  useEffect(() => {
    if (selectVariantEventRef.current) {
      if (typeof onSelect === 'function') {
        onSelect(enrichEventWithDetails(selectVariantEventRef.current, { selectedVariant }));
        selectVariantEventRef.current = undefined;
      }
    }
  }, [selectedVariant, onSelect]);

  const handleVariantItemSelect = (e) => {
    setSelectedVariant({ ...e.detail.selectedItems[0].dataset, variantItem: e.detail.selectedItems[0] });
    selectVariantEventRef.current = e;
    if (closeOnItemSelect) {
      handleClose();
    }
  };

  const variantNames = safeChildren.map((item: ComponentElement<any, any>) =>
    typeof item.props?.children === 'string' ? item.props.children : ''
  );

  const [favoriteChildren, setFavoriteChildren] = useState(undefined);

  useEffect(() => {
    if (showOnlyFavorites) {
      setFavoriteChildren(
        safeChildren.filter((child: ComponentElement<any, any>) => child.props.favorite || child.props.isDefault)
      );
    }
    if (!showOnlyFavorites && favoriteChildren?.length > 0) {
      setFavoriteChildren(undefined);
    }
  }, [showOnlyFavorites, safeChildren]);

  const safeChildrenWithFavorites = favoriteChildren ?? safeChildren;

  const [filteredChildren, setFilteredChildren] = useState(undefined);
  const [searchValue, setSearchValue] = useState('');
  const handleSearchInput = (e) => {
    setSearchValue(e.target.value);
    setFilteredChildren(
      safeChildrenWithFavorites.filter(
        (child: ComponentElement<any, any>) =>
          typeof child?.props?.children === 'string' &&
          child.props.children.toLowerCase().includes(e.target.value.toLowerCase())
      )
    );
  };
  useEffect(() => {
    if (filteredChildren) {
      setFilteredChildren(
        safeChildrenWithFavorites.filter(
          (child: ComponentElement<any, any>) =>
            typeof child?.props?.children === 'string' && child.props.children.toLowerCase().includes(searchValue)
        )
      );
    }
  }, [safeChildrenWithFavorites]);
  const handleSpaceInput = (e) => {
    if (e.code === 'Space') {
      setSearchValue((prev) => prev + ' ');
    }
  };
  const handleResetFilter = () => {
    setSearchValue('');
    setFilteredChildren(undefined);
  };

  const showSaveBtn = dirtyState && !selectedVariant?.readOnly;

  return (
    <div className={variantManagementClasses} style={style} title={tooltip} {...rest} ref={ref}>
      <VariantManagementContext.Provider
        value={{
          selectVariantItem: setSelectedVariant
        }}
      >
        <FlexBox onClick={disabled ? undefined : handleOpenVariantManagement}>
          <Title level={level} className={classes.title}>
            {selectedVariant?.children}
          </Title>
          {dirtyState && <div className={dirtyStateClasses}>{dirtyStateText}</div>}
        </FlexBox>
        <Button
          tooltip={selectViewText}
          aria-label={selectViewText}
          onClick={handleOpenVariantManagement}
          design={ButtonDesign.Transparent}
          icon="navigation-down-arrow"
          disabled={disabled}
        />
        {createPortal(
          <ResponsivePopover
            className={classes.popover}
            ref={popoverRef}
            headerText={titleText}
            placementType={placement}
            footer={
              (showSaveBtn || !hideSaveAs || !hideManageVariants || showCancelButton) && (
                <Bar
                  endContent={
                    <>
                      {!inErrorState && showSaveBtn && (
                        <Button onClick={handleSave} design={ButtonDesign.Emphasized}>
                          {saveText}
                        </Button>
                      )}
                      {!inErrorState && !hideSaveAs && (
                        <Button
                          onClick={handleOpenSaveAsDialog}
                          design={showSaveBtn ? ButtonDesign.Transparent : ButtonDesign.Emphasized}
                          disabled={!selectedVariant || Object.keys(selectedVariant).length === 0}
                        >
                          {saveAsText}
                        </Button>
                      )}
                      {!inErrorState && !hideManageVariants && (
                        <Button
                          onClick={handleManageClick}
                          design={showSaveBtn || !hideSaveAs ? ButtonDesign.Transparent : ButtonDesign.Emphasized}
                        >
                          {manageText}
                        </Button>
                      )}
                      {showCancelButton && (
                        <Button
                          onClick={handleClose}
                          design={
                            !inErrorState && (showSaveBtn || !hideSaveAs || !hideManageVariants)
                              ? ButtonDesign.Transparent
                              : ButtonDesign.Emphasized
                          }
                        >
                          {cancelText}
                        </Button>
                      )}
                    </>
                  }
                />
              )
            }
            onAfterClose={stopPropagation}
          >
            {inErrorState ? (
              <IllustratedMessage name={IllustrationMessageType.UnableToLoad} />
            ) : (
              <List
                onSelectionChange={handleVariantItemSelect}
                mode={ListMode.SingleSelect}
                header={
                  showInput ? (
                    <div className={classes.searchInput} tabIndex={-1}>
                      <Input
                        accessibleName={a11ySearchText}
                        value={searchValue}
                        placeholder={searchText}
                        onInput={handleSearchInput}
                        // todo remove when fixed
                        onKeyDown={handleSpaceInput}
                        icon={
                          <>
                            {filteredChildren && (
                              <Icon
                                accessibleName={resetIconTitleText}
                                tooltip={resetIconTitleText}
                                name="decline"
                                interactive
                                onClick={handleResetFilter}
                                className={classes.inputIcon}
                              />
                            )}
                            <Icon name="search" className={classes.inputIcon} />
                          </>
                        }
                      />
                    </div>
                  ) : undefined
                }
              >
                {filteredChildren ?? safeChildrenWithFavorites}
              </List>
            )}
          </ResponsivePopover>,
          portalContainer
        )}
        {manageViewsDialogOpen && (
          <ManageViewsDialog
            onAfterClose={handleManageClose}
            handleSaveManageViews={handleSaveManageViews}
            showShare={!hideShare}
            showApplyAutomatically={!hideApplyAutomatically}
            showSetAsDefault={!hideSetAsDefault}
            variantNames={variantNames}
            portalContainer={portalContainer}
          >
            {safeChildren}
          </ManageViewsDialog>
        )}
        {saveAsDialogOpen && (
          <SaveViewDialog
            showShare={!hideShare}
            showApplyAutomatically={!hideApplyAutomatically}
            showSetAsDefault={!hideSetAsDefault}
            onAfterClose={handleSaveAsClose}
            handleSave={handleSaveView}
            selectedVariant={selectedVariant}
            variantNames={variantNames}
          />
        )}
      </VariantManagementContext.Provider>
    </div>
  );
});

VariantManagement.defaultProps = {
  placement: PopoverPlacementType.Bottom,
  level: TitleLevel.H4,
  dirtyStateText: '*',
  portalContainer: document.body
};
VariantManagement.displayName = 'VariantManagement';

export { VariantManagement };
