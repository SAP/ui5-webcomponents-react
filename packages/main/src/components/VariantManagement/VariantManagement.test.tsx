import { fireEvent, getByText, render, renderWithDefine, screen, waitFor } from '@shared/tests';
import { createPassThroughPropsTest } from '@shared/tests/utils';
import { VariantManagement } from '@ui5/webcomponents-react/dist/VariantManagement';
import React from 'react';
import { TitleLevel } from '../..';
import { Ui5DialogDomRef } from '../../interfaces/Ui5DialogDomRef';
import { Ui5PopoverDomRef } from '../../interfaces/Ui5PopoverDomRef';
import { VariantItem } from './VariantItem';

const TwoVariantItems = [
  <VariantItem key="0">VariantItem 1</VariantItem>,
  <VariantItem selected key="1">
    VariantItem 2
  </VariantItem>
];

describe('VariantManagement', () => {
  test('Render without crashing', () => {
    const { rerender, getByTitle, getAllByText, getByText } = render(<VariantManagement />);
    getByTitle('Select view');
    //todo
    expect(screen).toMatchSnapshot();
    rerender(<VariantManagement>{TwoVariantItems}</VariantManagement>);
    getByTitle('Select view');
    expect(getAllByText('VariantItem 2')).toHaveLength(2);
    getByText('VariantItem 1');
    //todo
    // expect(document.body).toMatchSnapshot();
  });

  test('Selection', async () => {
    const cb = jest.fn((e) => e.detail);
    const { getByTitle, getAllByText, getByText } = await renderWithDefine(
      <VariantManagement onSelect={cb} showCancelButton>
        {TwoVariantItems}
      </VariantManagement>,
      ['ui5-li']
    );
    const wcListItem = screen.getByText('VariantItem 1');
    await waitFor(() => wcListItem.shadowRoot.querySelector('li'));
    const li = wcListItem.shadowRoot.querySelector('li');

    const popover: Ui5PopoverDomRef = document.querySelector('ui5-responsive-popover');
    const btn = getByTitle('Select view');
    const heading = getAllByText('VariantItem 2')[0];
    const closeBtn = getByText('Cancel');

    // initial selected Item
    expect(heading).toHaveTextContent('VariantItem 2');
    // open popover
    expect(popover.isOpen()).toBeFalsy();
    fireEvent.click(btn);
    expect(popover.isOpen()).toBeTruthy();
    // select list-item
    fireEvent.click(li);
    expect(cb).toHaveBeenCalledTimes(1);
    expect(getAllByText('VariantItem 1')).toHaveLength(2);
    expect(getAllByText('VariantItem 2')).toHaveLength(1);
    //todo
    // expect(document.body).toMatchSnapshot();
    expect(cb.mock.results[0].value.selectedVariant.children).toBe('VariantItem 1');
    expect(cb.mock.results[0].value.selectedVariant.variantItem).toBeInTheDocument();

    // close popover
    fireEvent.click(closeBtn);
    expect(popover.isOpen()).toBeFalsy();
    // open by clicking on heading
    fireEvent.click(getAllByText('VariantItem 1')[0]);
    expect(popover.isOpen()).toBeTruthy();
    fireEvent.click(closeBtn);
    expect(popover.isOpen()).toBeFalsy();
  });

  test('Disabled', () => {
    const { container, getByTitle } = render(<VariantManagement disabled>{TwoVariantItems}</VariantManagement>);

    const btn = getByTitle('Select view');
    const headingContainer = container.children[0];

    expect(headingContainer).toHaveClass('VariantManagement-disabled');
    expect(btn).toHaveAttribute('disabled', 'true');
    // todo
    // expect(document.body).toMatchSnapshot();
  });

  test('Close on item selection', async () => {
    const { getByTitle, rerender } = await renderWithDefine(
      <VariantManagement closeOnItemSelect>{TwoVariantItems}</VariantManagement>,
      ['ui5-li']
    );
    const wcListItem = screen.getByText('VariantItem 1');
    await waitFor(() => wcListItem.shadowRoot.querySelector('li'));
    const li = wcListItem.shadowRoot.querySelector('li');

    const popover: Ui5PopoverDomRef = document.querySelector('ui5-responsive-popover');
    const btn = getByTitle('Select view');
    fireEvent.click(btn);
    expect(popover.isOpen()).toBeTruthy();
    fireEvent.click(li);
    expect(popover.isOpen()).toBeFalsy();

    rerender(<VariantManagement>{TwoVariantItems}</VariantManagement>);
    fireEvent.click(btn);
    expect(popover.isOpen()).toBeTruthy();
    fireEvent.click(li);
    expect(popover.isOpen()).toBeTruthy();
  });

  test('Dirty state', async () => {
    const variantItems = [
      ...TwoVariantItems,
      <VariantItem key="2" readOnly>
        VariantItem 3
      </VariantItem>
    ];
    const cb = jest.fn((e) => e.detail);
    const { getByTitle, rerender, getByText, queryByText } = await renderWithDefine(
      <VariantManagement dirtyState onSave={cb}>
        {variantItems}
      </VariantManagement>,
      ['ui5-li']
    );
    const wcListItem = screen.getByText('VariantItem 3');
    await waitFor(() => wcListItem.shadowRoot.querySelector('li'));
    const li = wcListItem.shadowRoot.querySelector('li');

    getByText('*');
    const saveBtn = getByText('Save');
    fireEvent.click(saveBtn);
    expect(cb).toHaveBeenCalledTimes(1);
    expect(cb.mock.results[0].value.children).toBe('VariantItem 2');
    expect(cb.mock.results[0].value.variantItem).toBeInTheDocument();

    //todo
    // expect(document.body).toMatchSnapshot();

    const popover: Ui5PopoverDomRef = document.querySelector('ui5-responsive-popover');
    const btn = getByTitle('Select view');
    fireEvent.click(btn);
    expect(popover.isOpen()).toBeTruthy();
    fireEvent.click(li);
    expect(queryByText('Save')).toBeNull();

    rerender(
      <VariantManagement dirtyState dirtyStateText="Dirty state">
        {variantItems}
      </VariantManagement>
    );
    //todo
    // expect(document.body).toMatchSnapshot();
    getByText('Dirty state');
    rerender(<VariantManagement>{variantItems}</VariantManagement>);
    expect(queryByText('*')).toBeNull();
    expect(queryByText('Dirty state')).toBeNull();
  });

  test('Show/Hide buttons', () => {
    const { getByText, queryByText, rerender } = render(<VariantManagement>{TwoVariantItems}</VariantManagement>);
    getByText('Save As');
    getByText('Manage');

    rerender(<VariantManagement hideSaveAs>{TwoVariantItems}</VariantManagement>);
    getByText('Manage');
    expect(queryByText('Save As')).toBeNull();
    rerender(<VariantManagement hideManageVariants>{TwoVariantItems}</VariantManagement>);
    getByText('Save As');
    expect(queryByText('Manage')).toBeNull();
    rerender(<VariantManagement showCancelButton>{TwoVariantItems}</VariantManagement>);
    getByText('Cancel');
  });

  test('In error state', () => {
    render(<VariantManagement inErrorState>{TwoVariantItems}</VariantManagement>);
    expect(document.querySelector('ui5-illustrated-message')).toBeInTheDocument();
    // todo
    // expect(document.body).toMatchSnapshot();
  });

  test('Headings customization', async () => {
    const { getAllByText, getByTitle } = await renderWithDefine(
      <VariantManagement titleText="Popover Heading" level={TitleLevel.H1}>
        {TwoVariantItems}
      </VariantManagement>,
      ['ui5-responsive-popover']
    );
    const btn = getByTitle('Select view');
    fireEvent.click(btn);
    const popover: Ui5PopoverDomRef = document.querySelector('ui5-responsive-popover');
    await waitFor(() => popover.shadowRoot.querySelector('h2'));
    const popoverHeading = popover.shadowRoot.querySelector('h2');
    expect(popoverHeading).toHaveTextContent('Popover Heading');
    expect(getAllByText('VariantItem 2')[0]).toHaveAttribute('level', 'H1');
    //todo
    // expect(document.body).toMatchSnapshot();
  });

  test('Show search input', () => {
    const variantItems = [
      ...TwoVariantItems,
      ...new Array(8)
        .fill(':)')
        .map((_, index) => <VariantItem key={index + 3}>{`VariantItem ${index + 3}`}</VariantItem>)
    ];
    const { rerender, getByPlaceholderText, queryByPlaceholderText } = render(
      <VariantManagement>{variantItems}</VariantManagement>
    );
    const input = getByPlaceholderText('Search');
    fireEvent.input(input, { target: { value: 'VariantItem 10' } });
    screen.debug();
    expect(document.querySelectorAll('ui5-li')).toHaveLength(1);

    const [a, ...rest] = variantItems;
    rerender(<VariantManagement>{rest}</VariantManagement>);
    expect(queryByPlaceholderText('Search')).toBeNull();
    //todo snapshot
  });

  test('Show only favorite', () => {
    const variantItems = [
      <VariantItem key="0">VariantItem 1</VariantItem>,
      <VariantItem key="1">VariantItem 2</VariantItem>,
      <VariantItem favorite selected>
        Favorite VariantItem
      </VariantItem>,
      <VariantItem isDefault>Default VariantItem</VariantItem>
    ];
    const { rerender, queryByText, getAllByText, getByText } = render(
      <VariantManagement showOnlyFavorites>{variantItems}</VariantManagement>
    );
    expect(queryByText('VariantItem 1')).toBeNull();
    expect(queryByText('VariantItem 2')).toBeNull();
    expect(getAllByText('Favorite VariantItem')).toHaveLength(2);
    getByText('Default VariantItem');

    rerender(<VariantManagement>{variantItems}</VariantManagement>);
    getByText('VariantItem 1');
    getByText('VariantItem 2');
    expect(getAllByText('Favorite VariantItem')).toHaveLength(2);
    getByText('Default VariantItem');
    //todo snapshot
  });

  test('Save As', () => {
    const cb = jest.fn((e) => e.detail);
    const { getByText, queryByText } = render(<VariantManagement onSaveAs={cb}>{TwoVariantItems}</VariantManagement>);
    const saveAsBtn = getByText('Save As');

    fireEvent.click(saveAsBtn);
    const dialog: Ui5DialogDomRef = document.querySelector('ui5-dialog');
    expect(dialog).toHaveAttribute('header-text', 'Save View');
    expect(dialog.isOpen()).toBeTruthy();

    const inputField = document.querySelector('ui5-input');
    const checkboxes = document.querySelectorAll('ui5-checkbox');
    const saveBtn = getByText('Save');
    expect(inputField).toHaveValue('VariantItem 2');
    checkboxes.forEach((item: HTMLInputElement) => {
      expect(item.checked).toBeFalsy();
      fireEvent.change(item, { target: { checked: true } });
    });
    // invalid entries
    fireEvent.change(inputField, { target: { value: '' } });
    fireEvent.click(saveBtn);
    expect(dialog).toBeInTheDocument();
    expect(inputField).toHaveAttribute('value-state', 'Error');
    getByText('Please specify a view name');
    fireEvent.change(inputField, { target: { value: 'VariantItem 1' } });
    fireEvent.click(saveBtn);
    expect(dialog).toBeInTheDocument();
    expect(inputField).toHaveAttribute('value-state', 'Error');
    getByText('A file with this name already exists');

    // valid entry & save
    fireEvent.change(inputField, { target: { value: 'Updated!' } });
    expect(inputField).toHaveValue('Updated!');
    checkboxes.forEach((item: HTMLInputElement) => {
      expect(item.checked).toBeTruthy();
    });
    //todo snapshot
    fireEvent.click(saveBtn);
    expect(queryByText('Please specify a view name')).toBeNull();
    expect(queryByText('A file with this name already exists')).toBeNull();
    expect(cb).toHaveBeenCalledTimes(1);
    expect(dialog).not.toBeInTheDocument();
    expect(cb.mock.results[0].value.children).toBe('Updated!');
    expect(cb.mock.results[0].value.isDefault).toBeTruthy();
    expect(cb.mock.results[0].value.global).toBeTruthy();
    expect(cb.mock.results[0].value.applyAutomatically).toBeTruthy();
    expect(cb.mock.results[0].value.variantItem).toBeInTheDocument();
  });

  createPassThroughPropsTest(VariantManagement);
});
