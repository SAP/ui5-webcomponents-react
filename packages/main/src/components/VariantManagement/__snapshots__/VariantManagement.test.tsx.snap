// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VariantManagement Render without crashing 1`] = `
<body>
  Object {}
  <div>
    <div
      class="VariantManagement-container"
    >
      <div
        class="FlexBox-flexBox FlexBox-flexBoxDirectionRow FlexBox-justifyContentStart FlexBox-alignItemsStretch FlexBox-flexWrapNoWrap"
      >
        <ui5-title
          class="VariantManagement-title"
          level="H4"
          ui5-title=""
        />
      </div>
      <ui5-button
        aria-label="Select view"
        design="Transparent"
        icon="navigation-down-arrow"
        title="Select view"
        ui5-button=""
      />
    </div>
  </div>
  <ui5-responsive-popover
    class="VariantManagement-popover"
    header-text="My Views"
    horizontal-align="Center"
    placement-type="Bottom"
    ui5-responsive-popover=""
    vertical-align="Center"
  >
    <ui5-bar
      design="Header"
      slot="footer"
      ui5-bar=""
    >
      <ui5-button
        design="Emphasized"
        disabled="true"
        slot="endContent"
        ui5-button=""
      >
        Save As
      </ui5-button>
      <ui5-button
        design="Transparent"
        slot="endContent"
        ui5-button=""
      >
        Manage
      </ui5-button>
    </ui5-bar>
    <ui5-list
      accessible-role="listbox"
      busy-delay="1000"
      growing="None"
      mode="SingleSelect"
      separators="All"
      ui5-list=""
    />
  </ui5-responsive-popover>
</body>
`;
